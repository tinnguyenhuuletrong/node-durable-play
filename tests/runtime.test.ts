import { promisify } from "node:util";
import {
  ERuntimeMode,
  RuntimeContext,
  TraceLog,
} from "../src/libs/runtimeContext";
import { simpleError } from "./process/simple.error";
import { simpleSuccess } from "./process/simple.success";

describe("runtime - record", () => {
  it("simpleSuccess", async () => {
    const ctx = new RuntimeContext(ERuntimeMode.EREPLAY_AND_RUN);
    await ctx.run(simpleSuccess);
    const traces = ctx.getTraces();
    expect(traces).toMatchSnapshot();

    expect(ctx.doQuery("getCount", [])).toEqual(3);
  });

  it("simpleError", async () => {
    const ctx = new RuntimeContext(ERuntimeMode.EREPLAY_AND_RUN);
    try {
      await ctx.run(simpleError);
    } catch (error) {}
    const traces = ctx.getTraces();
    expect(traces).toMatchSnapshot();
    expect(ctx.doQuery("getCount", [])).toEqual(2);
  });
});

describe("runtime - replay", () => {
  it("simpleSuccess - replay only", async () => {
    const ctx = new RuntimeContext(ERuntimeMode.EREPLAY_ONLY);
    ctx.restore([
      {
        action: "doInc",
        error: undefined,
        isSuccess: true,
        params: [],
        result: 1,
        seqId: 1,
      },
      {
        action: "waitMs",
        error: undefined,
        isSuccess: true,
        params: [1000],
        result: undefined,
        seqId: 2,
      },
    ]);
    await ctx.replay(simpleSuccess);

    expect(ctx.doQuery("getCount", [])).toEqual(1);
  });

  it("simpleSuccess - replay and continue", async () => {
    const ctx = new RuntimeContext(ERuntimeMode.EREPLAY_AND_RUN);
    ctx.restore([
      {
        action: "doInc",
        error: undefined,
        isSuccess: true,
        params: [],
        result: 1,
        seqId: 1,
      },
      {
        action: "waitMs",
        error: undefined,
        isSuccess: true,
        params: [1000],
        result: undefined,
        seqId: 2,
      },
    ]);
    await ctx.run(simpleSuccess);

    expect(ctx.doQuery("getCount", [])).toEqual(3);
  });
});
