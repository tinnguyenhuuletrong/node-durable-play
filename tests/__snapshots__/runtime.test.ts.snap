// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime - record simpleError 1`] = `
[
  {
    "action": "doInc",
    "error": undefined,
    "isSuccess": true,
    "params": [],
    "result": 1,
    "seqId": 1,
  },
  {
    "action": "waitMs",
    "error": undefined,
    "isSuccess": true,
    "params": [
      1000,
    ],
    "result": undefined,
    "seqId": 2,
  },
  {
    "action": "doInc",
    "error": [Error: ðŸ˜ª],
    "isSuccess": false,
    "params": [],
    "result": undefined,
    "seqId": 3,
  },
]
`;

exports[`runtime - record simpleSuccess 1`] = `
[
  {
    "action": "doInc",
    "error": undefined,
    "isSuccess": true,
    "params": [],
    "result": 1,
    "seqId": 1,
  },
  {
    "action": "waitMs",
    "error": undefined,
    "isSuccess": true,
    "params": [
      1000,
    ],
    "result": undefined,
    "seqId": 2,
  },
  {
    "action": "doInc",
    "error": undefined,
    "isSuccess": true,
    "params": [],
    "result": 2,
    "seqId": 3,
  },
  {
    "action": "waitMs",
    "error": undefined,
    "isSuccess": true,
    "params": [
      1000,
    ],
    "result": undefined,
    "seqId": 4,
  },
  {
    "action": "doInc",
    "error": undefined,
    "isSuccess": true,
    "params": [],
    "result": 3,
    "seqId": 5,
  },
  {
    "action": "waitMs",
    "error": undefined,
    "isSuccess": true,
    "params": [
      1000,
    ],
    "result": undefined,
    "seqId": 6,
  },
]
`;
